pkg/trainer/tensorboard.go

```diff
@@ -0,0 +1,196 @@
+}
+
+func (s *TBReplicaSet) Create() error {
+	//By default we assume TensorBoard's service will be a ClusterIP
+	// unless specified otherwise by the user
+	st := v1.ServiceType("ClusterIP")
+	if s.Spec.ServiceType != "" {
+		st = s.Spec.ServiceType
+	}
+
+	//create the service exposing TensorBoard
+	service := &v1.Service{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name:   s.jobName(),
+			Labels: s.Labels(),
+		},
+		Spec: v1.ServiceSpec{
+			Type:     st,
+			Selector: s.Labels(),
+			Ports: []v1.ServicePort{
+				{
+					Name: "tb-port",
+					Port: 80,
+					TargetPort: intstr.IntOrString{
+						IntVal: 6006,
+					},
+				},
+			},
+		},
+	}
+
+	log.Infof("Creating Service: %v", service.ObjectMeta.Name)
+	_, err := s.ClientSet.CoreV1().Services(NAMESPACE).Create(service)
+
+	// If the job already exists do nothing.
+	if err != nil {
+		if k8s_errors.IsAlreadyExists(err) {
+			log.Infof("Service %v already exists.", s.jobName())
+		} else {
+			return err
+		}
+	}
+
+	newD := &v1beta1.Deployment{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name:   s.jobName(),
+			Labels: s.Labels(),
+		},
+		Spec: v1beta1.DeploymentSpec{
+			Selector: &meta_v1.LabelSelector{
+				MatchLabels: s.Labels(),
+			},
+			Replicas: proto.Int32(1),
+			Template: s.getDeploymentSpecTemplate(),
+		},
+	}
+
+	log.Infof("Creating Deployment: %v", newD.ObjectMeta.Name)
+	_, err = s.ClientSet.ExtensionsV1beta1().Deployments(NAMESPACE).Create(newD)
+
+	if err != nil {
+		if k8s_errors.IsAlreadyExists(err) {
+			log.Infof("%v already exists.", s.jobName())
+		} else {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *TBReplicaSet) Delete() error {
+	selector, err := s.Labels().ToSelector()
+	if err != nil {
+		return err
+	}
+
+	listOpts := meta_v1.ListOptions{
+		LabelSelector: selector,
+	}
+
+	failures := false
+
+	err = s.ClientSet.ExtensionsV1beta1().Deployments(NAMESPACE).Delete(s.jobName(), &meta_v1.DeleteOptions{})
+	if err != nil {
+		log.Errorf("There was a problem deleting TensorBoard's deployment %v; %v", s.jobName(), err)
+		failures = true
+	}
+
+	err = s.ClientSet.ExtensionsV1beta1().ReplicaSets(NAMESPACE).DeleteCollection(&meta_v1.DeleteOptions{}, listOpts)
+	if err != nil {
+		log.Errorf("Error deleting TensorBoard's replica set: %v; %v", s.jobName(), err)
+		failures = true
+	}
+
+	err = s.ClientSet.CoreV1().Pods(NAMESPACE).DeleteCollection(&meta_v1.DeleteOptions{}, listOpts)
+	if err != nil {
+		log.Errorf("Error deleting TensorBoard's pod: %v; %v", s.jobName(), err)
+		failures = true
+	}
+
+	err = s.ClientSet.CoreV1().Services(NAMESPACE).Delete(s.jobName(), &meta_v1.DeleteOptions{})
+	if err != nil {
+		log.Errorf("Error deleting service: %v; %v", s.jobName(), err)
+		failures = true
+	}
+
+	if failures {
+		return errors.New("There was an issue deleting TensorBoard's resources")
+	}
+	return nil
+}
+
+func (s *TBReplicaSet) getDeploymentSpecTemplate() v1.PodTemplateSpec {
+	c := &v1.Container{
+		Name:  s.jobName(),
+		Image: "tensorflow/tensorflow",
```

Comment:
Maybe add a TODO here to make the TensorFlow image a parameter of the job operator. I think that there's a couple places where we might want to just use the default TensorFlow image. For example in parameter servers we can just start the TensorFlow server.

However, users might want to specify which version of TensorFlow they want to use e.g. 1.0/1.2 etc...